### PR TITLE
basic spectra analysis

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR adds a basic spectra analysis relying on chi^2 minimization (similar to sncosmo for light curves).

### CODE REVIEW

File: data/db_demo.yaml [hunk 1]
```diff
@@ -1055,3 +1055,17 @@ analysis_service:
     group_ids:
       - =program_A
       - =program_B
+  - name: "NGSF_Analysis"
+    display_name: "NGSF analysis"
+    description: "Use NGSF to fit spectra"
```
Comment (guynir42) [inline data/db_demo.yaml:1060]: Spell out the acronym please!

File: services/ngsf_analysis_service/app.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    try:
+        _ = requests.post(
+            url,
+            json=results,
+            timeout=request_timeout,
+        )
+    except requests.exceptions.Timeout:
+        # If we timeout here then it's precisely because
+        # we cannot write back to the SkyPortal instance.
+        # So returning something doengsf't make sense in this case.
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:57]:
```suggestion
        # So returning something doesn't make sense in this case.
```

File: services/ngsf_analysis_service/app.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        # Just log it and move on...
+        log("Callback URL timedout. Skipping.")
+    except Exception as e:
+        log(f"Callback exception {e}.")
+
+
+def run_ngsf_model(data_dict):
+    """
+    Use `ngsf` to fit data to a model with name `source_name`.
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:66]: Spell it out here as well. Thanks!

File: services/ngsf_analysis_service/app.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    For this analysis, we expect the `inputs` dictionary to have the following keys:
+       - source: the name of the model to fit to the data
+       - fix_z: whether to fix the redshift
+       - photometry: the photometry to fit to the model (in csv format)
+       - redshift: the known redshift of the object
+
+    Other analysis services may require additional keys in the `inputs` dictionary.
+    """
+    analysis_parameters = data_dict["inputs"].get("analysis_parameters", {})
+    analysis_parameters = {**default_analysis_parameters, **analysis_parameters}
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:77]:
```suggestion
    analysis_parameters.update(default_analysis_parameters)
```

Comment (guynir42) [reply services/ngsf_analysis_service/app.py:77]: In python 3.9 you can also do `analysis_parameters = default_analysis_parameters | analysis_parameters` though I'm not sure it is better and we are on 3.8 anyway.

File: services/ngsf_analysis_service/app.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    # this example analysis service expects the photometry to be in
+    # a csv file (at data_dict["inputs"]["photometry"]) with the following columns
+    # - filter: the name of the bandpass
+    # - mjd: the modified Julian date of the observation
+    # - magsys: the mag system (e.g. ab) of the observations
+    # - flux: the flux of the observation
+    #
+    # the following code transforms these inputs from SkyPortal
+    # to the format expected by ngsfcosmo.
+    #
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:90]: May need to update this comment.

File: services/ngsf_analysis_service/app.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+    local_temp_files = []
+    plot_data = []
+
+    try:
+        for index, row in data.iterrows():
+            filebase = str(uuid.uuid4())
+            SPECFILE = f'{SUPERFIT_DATA_PATH}/{filebase}.dat'
+            wavelengths = np.array(ast.literal_eval(row['wavelengths']))
+            fluxes = np.array(ast.literal_eval(row['fluxes']))
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:140]: This seems like a strange way to read a file. Any idea what is happening here?

File: services/ngsf_analysis_service/app.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+            fid = open(SPECFILE, 'w')
+            for w, f in zip(wavelengths.tolist(), fluxes.tolist()):
+                fid.write(f'{w} {f}\n')
+            fid.close()
```
Comment (guynir42) [inline services/ngsf_analysis_service/app.py:144]:
```suggestion
            with open(SPECFILE, 'w') as fid:
                for w, f in zip(wavelengths.tolist(), fluxes.tolist()):
                    fid.write(f'{w} {f}\n')
```

**Review by guynir42 — APPROVED** [review]
This is pretty cool. I actually helped optimize the python superfit a couple of years ago. It would be nice to see it in action on fritz.

File: services/ngsf_analysis_service/app.py (hunk 7 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py]: so fancy.

File: services/ngsf_analysis_service/app.py (hunk 2 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py]: good catch.

File: services/ngsf_analysis_service/app.py (hunk 3 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py]: Yup.

File: services/ngsf_analysis_service/app.py (hunk 4 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py:77]: Someday we will make it to 3.9

File: services/ngsf_analysis_service/app.py (hunk 5 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py]: Good call ;)

File: services/ngsf_analysis_service/app.py (hunk 6 above)
Comment (mcoughlin) [reply services/ngsf_analysis_service/app.py:138]: It's sent as a list (including the []) from the serializer, so its the easiest way to pull this in.

**Pull request merged**